README.rst
LSHash
======

:Version: 0.0.8
:Python: 3.7.7

A fast Python implementation of locality sensitive hashing with persistance
support.

Based on original source code https://github.com/kayzhu/LSHash

Highlights
==========

- Python3 support
- Load & save hash tables to local disk
- Fast hash calculation for large amount of high dimensional data through the use of `numpy` arrays.
- Built-in support for persistency through Redis.
- Multiple hash indexes support.
- Built-in support for common distance/objective functions for ranking outputs.

Installation
============
``LSHash`` depends on the following libraries:

- numpy
- bitarray (if hamming distance is used as distance function)

Optional
- redis (if persistency through Redis is needed)

To install from sources:

.. code-block:: bash

    $ git clone https://github.com/loretoparisi/lshash.git
    $ python setup.py install
    
To install from PyPI:

.. code-block:: bash

    $ pip install lshashpy3
    $ python -c "import lshashpy3 as lshash; print(lshash.__version__);"

Quickstart
==========
To create 6-bit hashes for input data of 8 dimensions:

.. code-block:: python

 # create 6-bit hashes for input data of 8 dimensions:
 lsh = LSHash(6, 8)
 
 # index vector
 lsh.index([2,3,4,5,6,7,8,9])

 # index vector and extra data
 lsh.index([10,12,99,1,5,31,2,3], extra_data="vec1")
 lsh.index([10,11,94,1,4,31,2,3], extra_data="vec2")

 # query a data point
 top_n = 1
 nn = lsh.query([1,2,3,4,5,6,7,7], num_results=top_n, distance_func="euclidean")
 print(nn)

 # unpack vector, extra data and vectorial distance
 top_n = 3
 nn = lsh.query([10,12,99,1,5,30,1,1], num_results=top_n, distance_func="euclidean")
    for ((vec,extra_data),distance) in nn:
        print(vec, extra_data, distance)
        
        
To save hash table to disk:

.. code-block:: python

 lsh = LSHash(hash_size=k, input_dim=d, num_hashtables=L,
     storage_config={ 'dict': None },
     matrices_filename='weights.npz', 
     hashtable_filename='hash.npz', 
     overwrite=True)

 lsh.index([10,12,99,1,5,31,2,3], extra_data="vec1")
 lsh.index([10,11,94,1,4,31,2,3], extra_data="vec2")
 lsh.save()

To load hash table from disk and perform a query:

.. code-block:: python

 lsh = LSHash(hash_size=k, input_dim=d, num_hashtables=L,
     storage_config={ 'dict': None },
     matrices_filename='weights.npz', 
     hashtable_filename='hash.npz', 
     overwrite=True)

 top_n = 3
 nn = lsh.query([10,12,99,1,5,30,1,1], num_results=top_n, distance_func="euclidean")
 print(nn)

API
==============

- To initialize a ``LSHash`` instance:

.. code-block:: python

 k = 6 # hash size
 L = 5  # number of tables
 d = 8 # Dimension of Feature vector
 LSHash(hash_size=k, input_dim=d, num_hashtables=L,
    storage_config={ 'dict': None },
    matrices_filename='weights.npz', 
    hashtable_filename='hash.npz', 
    overwrite=True)

parameters:

``hash_size``:
    The length of the resulting binary hash.
``input_dim``:
    The dimension of the input vector.
``num_hashtables = 1``:
    (optional) The number of hash tables used for multiple lookups.
``storage = None``:
    (optional) Specify the name of the storage to be used for the index
    storage. Options include "redis".
``matrices_filename = None``:
    (optional) Specify the path to the .npz file random matrices are stored
    or to be stored if the file does not exist yet
``hashtable_filename = None``:
    (optional) Specify the path to the .npz file hash table are stored
    or to be stored if the file does not exist yet
``overwrite = False``:
    (optional) Whether to overwrite the matrices file if it already exist

- To index a data point of a given ``LSHash`` instance, e.g., ``lsh``:

.. code-block:: python

    lsh.index(input_point, extra_data=None):

parameters:

``input_point``:
    The input data point is an array or tuple of numbers of input_dim.
``extra_data = None``:
    (optional) Extra data to be added along with the input_point.

- To query a data point against a given ``LSHash`` instance, e.g., ``lsh``:

.. code-block:: python

    lsh.query(query_point, num_results=None, distance_func="euclidean"):

parameters:

``query_point``:
    The query data point is an array or tuple of numbers of input_dim.
``num_results = None``:
    (optional) The number of query results to return in ranked order. By
    default all results will be returned.
``distance_func = "euclidean"``:
    (optional) Distance function to use to rank the candidates. By default
    "euclidean" distance function will be used. Distance function can be 
    "euclidean", "true_euclidean", "centred_euclidean", "cosine", "l1norm".
    

- To save the hash table currently indexed:

.. code-block:: python

    lsh.save():


CHANGES.rst
:Version: 0.0.8
:Python: 3.7.7

lshashpy3/lshash.py
# lshash/lshash.py
# Copyright 2012 Kay Zhu (a.k.a He Zhu) and contributors (see CONTRIBUTORS.txt)
#
# This module is part of lshash and is released under
# the MIT License: http://www.opensource.org/licenses/mit-license.php
# -*- coding: utf-8 -*-
from __future__ import print_function, unicode_literals, division, absolute_import
from builtins import int, round, str,  object  # noqa
from future import standard_library
standard_library.install_aliases()  # noqa: Counter, OrderedDict, 
from past.builtins import basestring   # noqa:
 
import future        # noqa
import builtins      # noqa
import past          # noqa
import six           # noqa
 
import os
import json
import numpy as np
 
try:
    from storage import storage  # py2
except ImportError:
    from .storage import storage  # py3
 
try:
    from bitarray import bitarray
except ImportError:
    bitarray = None
 
 
try:
    xrange  # py2
except NameError:
    xrange = range  # py3
 
 
class LSHash(object):
    """ LSHash implments locality sensitive hashing using random projection for
    input vectors of dimension `input_dim`.
 
    Attributes:
 
    :param hash_size:
        The length of the resulting binary hash in integer. E.g., 32 means the
        resulting binary hash will be 32-bit long.
    :param input_dim:
        The dimension of the input vector. E.g., a grey-scale picture of 30x30
        pixels will have an input dimension of 900.
    :param num_hashtables:
        (optional) The number of hash tables used for multiple lookups.
    :param storage_config:
        (optional) A dictionary of the form `{backend_name: config}` where
        `backend_name` is the either `dict` or `redis`, and `config` is the
        configuration used by the backend. For `redis` it should be in the
        format of `{"redis": {"host": hostname, "port": port_num}}`, where
        `hostname` is normally `localhost` and `port` is normally 6379.
    :param matrices_filename:
        (optional) Specify the path to the compressed numpy file ending with
        extension `.npz`, where the uniform random planes are stored, or to be
        stored if the file does not exist yet.
    :param overwrite:
        (optional) Whether to overwrite the matrices file if it already exist
    """
 
    def __init__(self, hash_size, input_dim, num_hashtables=1,
                 storage_config=None, matrices_filename=None, hashtable_filename=None, overwrite=False):
 
        self.hash_size = hash_size
        self.input_dim = input_dim
        self.num_hashtables = num_hashtables
 
        if storage_config is None:
            storage_config = {'dict': None}
        self.storage_config = storage_config
 
        if matrices_filename and not matrices_filename.endswith('.npz'):
            raise ValueError("The specified file name must end with .npz")
        self.matrices_filename = matrices_filename

        if hashtable_filename and not hashtable_filename.endswith('.npz'):
            raise ValueError("The specified file name must end with .npz")
        self.hashtable_filename = hashtable_filename

        self.overwrite = overwrite
 
        self._init_uniform_planes()
        self._init_hashtables()
 
    def _init_uniform_planes(self):
        """ Initialize uniform planes used to calculate the hashes
 
        if file `self.matrices_filename` exist and `self.overwrite` is
        selected, save the uniform planes to the specified file.
 
        if file `self.matrices_filename` exist and `self.overwrite` is not
        selected, load the matrix with `np.load`.
 
        if file `self.matrices_filename` does not exist and regardless of
        `self.overwrite`, only set `self.uniform_planes`.
        """
 
        if "uniform_planes" in self.__dict__:
            return
 
        if self.matrices_filename:
            file_exist = os.path.isfile(self.matrices_filename)
            if file_exist and not self.overwrite:
                try:
                    npzfiles = np.load(self.matrices_filename)
                except IOError:
                    print("Cannot load specified file as a numpy array")
                    raise
                else:
                    npzfiles = sorted(npzfiles.items(), key=lambda x: x[0])
                    self.uniform_planes = [t[1] for t in npzfiles]
            else:
                self.uniform_planes = [self._generate_uniform_planes()
                                       for _ in xrange(self.num_hashtables)]
                try:
                    np.savez_compressed(self.matrices_filename,
                                        *self.uniform_planes)
                except IOError:
                    print("IOError when saving matrices to specificed path")
                    raise
        else:
            self.uniform_planes = [self._generate_uniform_planes()
                                   for _ in xrange(self.num_hashtables)]
 
    def _init_hashtables(self):
        """ Initialize the hash tables such that each record will be in the
        form of "[storage1, storage2, ...]" """

        if self.hashtable_filename:
            file_exist = os.path.isfile(self.hashtable_filename)
            if file_exist:
                try:
                    npzfiles = np.load(self.hashtable_filename, allow_pickle=True)
                    self.hash_tables = npzfiles['data']
                except IOError:
                    print("Cannot load specified file as a numpy array")
                    raise
            else:
                self.hash_tables = [storage(self.storage_config, i)
                                for i in xrange(self.num_hashtables)]
        else:
            self.hash_tables = [storage(self.storage_config, i)
                                for i in xrange(self.num_hashtables)]
 
    def _generate_uniform_planes(self):
        """ Generate uniformly distributed hyperplanes and return it as a 2D
        numpy array.
        """
 
        return np.random.randn(self.hash_size, self.input_dim)
 
    def _hash(self, planes, input_point):
        """ Generates the binary hash for `input_point` and returns it.
 
        :param planes:
            The planes are random uniform planes with a dimension of
            `hash_size` * `input_dim`.
        :param input_point:
            A Python tuple or list object that contains only numbers.
            The dimension needs to be 1 * `input_dim`.
        """
 
        try:
            input_point = np.array(input_point)  # for faster dot product
            projections = np.dot(planes, input_point)
        except TypeError as e:
            print("""The input point needs to be an array-like object with
                  numbers only elements""")
            raise
        except ValueError as e:
            print("""The input point needs to be of the same dimension as
                  `input_dim` when initializing this LSHash instance""", e)
            raise
        else:
            return "".join(['1' if i > 0 else '0' for i in projections])
 
    def _as_np_array(self, json_or_tuple):
        """ Takes either a JSON-serialized data structure or a tuple that has
        the original input points stored, and returns the original input point
        in numpy array format.
        """
        if isinstance(json_or_tuple, basestring):
            # JSON-serialized in the case of Redis
            try:
                # Return the point stored as list, without the extra data
                tuples = json.loads(json_or_tuple)[0]
            except TypeError:
                print("The value stored is not JSON-serilizable")
                raise
        else:
            # If extra_data exists, `tuples` is the entire
            # (point:tuple, extra_data). Otherwise (i.e., extra_data=None),
            # return the point stored as a tuple
            tuples = json_or_tuple
 
        if isinstance(tuples[0], tuple):
            # in this case extra data exists
            return np.asarray(tuples[0])
 
        elif isinstance(tuples, (tuple, list)):
            try:
                return np.asarray(tuples)
            except ValueError as e:
                print("The input needs to be an array-like object", e)
                raise
        else:
            raise TypeError("query data is not supported")
 
    def index(self, input_point, extra_data=None):
        """ Index a single input point by adding it to the selected storage.
 
        If `extra_data` is provided, it will become the value of the dictionary
        {input_point: extra_data}, which in turn will become the value of the
        hash table. `extra_data` needs to be JSON serializable if in-memory
        dict is not used as storage.
 
        :param input_point:
            A list, or tuple, or numpy ndarray object that contains numbers
            only. The dimension needs to be 1 * `input_dim`.
            This object will be converted to Python tuple and stored in the
            selected storage.
        :param extra_data:
            (optional) Needs to be a JSON-serializable object: list, dicts and
            basic types such as strings and integers.
        """
 
        if isinstance(input_point, np.ndarray):
            input_point = input_point.tolist()
 
        if extra_data:
            value = (tuple(input_point), extra_data)
        else:
            # LP: extra_data to None to keep output consistency
            value = (tuple(input_point), None)
 
        for i, table in enumerate(self.hash_tables):
            table.append_val(self._hash(self.uniform_planes[i], input_point),
                             value)
    
    def save(self):
        """
            Save save the uniform planes to the specified file.
        """
        if self.hashtable_filename:
            try:
                np.savez_compressed(self.hashtable_filename, allow_pickle=True, data=self.hash_tables)

            except IOError:
                print("IOError when saving hash tables to specificed path")
                raise

    def query(self, query_point, num_results=None, distance_func=None):
        """ Takes `query_point` which is either a tuple or a list of numbers,
        returns `num_results` of results as a list of tuples that are ranked
        based on the supplied metric function `distance_func`.
 
        :param query_point:
            A list, or tuple, or numpy ndarray that only contains numbers.
            The dimension needs to be 1 * `input_dim`.
            Used by :meth:`._hash`.
        :param num_results:
            (optional) Integer, specifies the max amount of results to be
            returned. If not specified all candidates will be returned as a
            list in ranked order.
        :param distance_func:
            (optional) The distance function to be used. Currently it needs to
            be one of ("hamming", "euclidean", "true_euclidean",
            "centred_euclidean", "cosine", "l1norm"). By default "euclidean"
            will used.
        """
 
        candidates = set()
        if not distance_func:
            distance_func = "euclidean"
 
        if distance_func == "hamming":
            if not bitarray:
                raise ImportError(" Bitarray is required for hamming distance")
 
            for i, table in enumerate(self.hash_tables):
                binary_hash = self._hash(self.uniform_planes[i], query_point)
                for key in table.keys():
                    distance = LSHash.hamming_dist(key, binary_hash)
                    if distance < 2:
                        candidates.update(table.get_list(key))
 
            d_func = LSHash.euclidean_dist_square
 
        else:
 
            if distance_func == "euclidean":
                d_func = LSHash.euclidean_dist_square
            elif distance_func == "true_euclidean":
                d_func = LSHash.euclidean_dist
            elif distance_func == "centred_euclidean":
                d_func = LSHash.euclidean_dist_centred
            elif distance_func == "cosine":
                d_func = LSHash.cosine_dist
            elif distance_func == "l1norm":
                d_func = LSHash.l1norm_dist
            else:
                raise ValueError("The distance function name is invalid.")
 
            for i, table in enumerate(self.hash_tables):
                binary_hash = self._hash(self.uniform_planes[i], query_point)
                candidates.update(table.get_list(binary_hash))
 
        # rank candidates by distance function
        candidates = [(ix, d_func(query_point, self._as_np_array(ix)))
                      for ix in candidates]
        candidates = sorted(candidates, key=lambda x: x[1])
 
        return candidates[:num_results] if num_results else candidates
 
    ### distance functions
 
    @staticmethod
    def hamming_dist(bitarray1, bitarray2):
        xor_result = bitarray(bitarray1) ^ bitarray(bitarray2)
        return xor_result.count()
 
    @staticmethod
    def euclidean_dist(x, y):
        """ This is a hot function, hence some optimizations are made. """
        diff = np.array(x) - y
        return np.sqrt(np.dot(diff, diff))
 
    @staticmethod
    def euclidean_dist_square(x, y):
        """ This is a hot function, hence some optimizations are made. """
        diff = np.array(x) - y
        return np.dot(diff, diff)
 
    @staticmethod
    def euclidean_dist_centred(x, y):
        """ This is a hot function, hence some optimizations are made. """
        diff = np.mean(x) - np.mean(y)
        return np.dot(diff, diff)
 
    @staticmethod
    def l1norm_dist(x, y):
        return sum(abs(x - y))
 
    @staticmethod
    def cosine_dist(x, y):
        return 1 - float(np.dot(x, y)) / ((np.dot(x, x) * np.dot(y, y)) ** 0.5)

lshashpy3/storage.py
# lshash/storage.py
# Copyright 2012 Kay Zhu (a.k.a He Zhu) and contributors (see CONTRIBUTORS.txt)
#
# This module is part of lshash and is released under
# the MIT License: http://www.opensource.org/licenses/mit-license.php
 
import json
 
try:
    import redis
except ImportError:
    redis = None
 
__all__ = ['storage']
 
 
def storage(storage_config, index):
    """ Given the configuration for storage and the index, return the
    configured storage instance.
    """
    if 'dict' in storage_config:
        return InMemoryStorage(storage_config['dict'])
    elif 'redis' in storage_config:
        storage_config['redis']['db'] = index
        return RedisStorage(storage_config['redis'])
    else:
        raise ValueError("Only in-memory dictionary and Redis are supported.")
 
 
class BaseStorage(object):
    def __init__(self, config):
        """ An abstract class used as an adapter for storages. """
        raise NotImplementedError
 
    def keys(self):
        """ Returns a list of binary hashes that are used as dict keys. """
        raise NotImplementedError
 
    def set_val(self, key, val):
        """ Set `val` at `key`, note that the `val` must be a string. """
        raise NotImplementedError
 
    def get_val(self, key):
        """ Return `val` at `key`, note that the `val` must be a string. """
        raise NotImplementedError
 
    def append_val(self, key, val):
        """ Append `val` to the list stored at `key`.
 
        If the key is not yet present in storage, create a list with `val` at
        `key`.
        """
        raise NotImplementedError
 
    def get_list(self, key):
        """ Returns a list stored in storage at `key`.
 
        This method should return a list of values stored at `key`. `[]` should
        be returned if the list is empty or if `key` is not present in storage.
        """
        raise NotImplementedError
 
 
class InMemoryStorage(BaseStorage):
    def __init__(self, config):
        self.name = 'dict'
        self.storage = dict()
 
    def keys(self):
        return self.storage.keys()
 
    def set_val(self, key, val):
        self.storage[key] = val
 
    def get_val(self, key):
        return self.storage[key]
 
    def append_val(self, key, val):
        self.storage.setdefault(key, []).append(val)
 
    def get_list(self, key):
        return self.storage.get(key, [])
 
 
class RedisStorage(BaseStorage):
    def __init__(self, config):
        if not redis:
            raise ImportError("redis-py is required to use Redis as storage.")
        self.name = 'redis'
        self.storage = redis.StrictRedis(**config)
 
    def keys(self, pattern="*"):
        return self.storage.keys(pattern)
 
    def set_val(self, key, val):
        self.storage.set(key, val)
 
    def get_val(self, key):
        return self.storage.get(key)
 
    def append_val(self, key, val):
        self.storage.rpush(key, json.dumps(val))
 
    def get_list(self, key):
        return self.storage.lrange(key, 0, -1)

example.py
# -*- coding: utf-8 -*-

from lshashpy3 import LSHash

# create 6-bit hashes for input data of 8 dimensions:
k = 6 # hash size
L = 5  # number of tables
d = 8 # Dimension of Feature vector
lsh = LSHash(hash_size=k, input_dim=d, num_hashtables=L)

# index vector
lsh.index([2,3,4,5,6,7,8,9])

# index vector and extra data
lsh.index([10,12,99,1,5,31,2,3], extra_data="vec1")
lsh.index([10,11,94,1,4,31,2,3], extra_data="vec2")

# query a data point
top_n = 1
nn = lsh.query([1,2,3,4,5,6,7,7], num_results=top_n, distance_func="euclidean")
print("query (euclidean):", nn)

# query a data point
top_n = 1
nn = lsh.query([1,2,3,4,5,6,7,7], num_results=top_n, distance_func="hamming")
print("query (hamming):", nn)

# unpack vector, extra data and vectorial distance
# distance_func can be "euclidean", "true_euclidean", "centred_euclidean", "cosine", "l1norm".
top_n = 3
nn = lsh.query([10,12,99,1,5,30,1,1], num_results=top_n, distance_func="euclidean")
for ((vec,extra_data),distance) in nn:
    print(vec, extra_data, distance)


# InMemoryStorage
lsh = LSHash(hash_size=k, input_dim=d, num_hashtables=L,
    storage_config={ 'dict': None }, matrices_filename='weights.npz', overwrite=False)

# local storage for numpy uniform random planes, overwrite matrix file
lsh = LSHash(hash_size=k, input_dim=d, num_hashtables=L,
    storage_config={ 'dict': None },
    matrices_filename='weights.npz', 
    hashtable_filename='hash.npz', 
    overwrite=True)

lsh.index([10,12,99,1,5,31,2,3], extra_data="vec1")
lsh.index([10,11,94,1,4,31,2,3], extra_data="vec2")

top_n = 3
nn = lsh.query([10,12,99,1,5,30,1,1], num_results=top_n, distance_func="euclidean")
print("query (euclidean):", nn)

# save hash table to disk 
lsh.save()

# local storage for numpy uniform random planes, overwrite matrix file
lsh = LSHash(hash_size=k, input_dim=d, num_hashtables=L,
    storage_config={ 'dict': None },
    matrices_filename='weights.npz', 
    hashtable_filename='hash.npz', 
    overwrite=True)

# execute a query loading hash table from local file system
top_n = 3
nn = lsh.query([10,12,99,1,5,30,1,1], num_results=top_n, distance_func="euclidean")
print("query from disk (euclidean):", nn)